File: static_deque.py
class StaticDeque:

    def __init__(self, max_len: int):
        self.max_len = max_len
        self.deque = [None] * self.max_len
        self.head = 0
        self.tail = 0

    def is_full(self) -> bool:
        return self.deque[self.tail] is not None

    def is_empty(self) -> bool:
        for el in self.deque:
            if el is not None:
                return False
        else:
            return True

    def appendleft(self, x) -> str:
        if self.is_full():
            print(f"Full deque, cannot append {x} left")
        else:
            self.deque[self.tail] = x
            self.tail = (self.tail + 1) % self.max_len
            return f"Element {x} added at position {self.tail}"

    def append(self, x) -> str:
        if self.is_full():
            print(f"Full deque, cannot append {x}")
        else:
            self.head = (self.head - 1) % self.max_len
            self.deque[self.head] = x
            return f"Element {x} added at position {self.head}"

    def clear(self) -> str:
        if self.deque != [None] * self.max_len:
            self.head = 0
            self.tail = 0
            self.deque = [None] * self.max_len
            return "Deque cleared"
        else:
            return "Deque is already cleared"

    def copy(self):
        new_deque = StaticDeque(self.max_len)
        new_deque.deque = self.deque.copy()
        new_deque.head = self.head
        new_deque.tail = self.tail
        return new_deque

    def count(self, x):
        return f"Item {x} is found {self.deque.count(x)} time(s)"

    def insert(self, i: int, x):
        if not (0 <= i < self.max_len):
            print("Index out of range")
        elif self.deque[i] is None:
            self.deque[i] = x
        else:
            print(f"There's already value ({self.deque[i]}) at index {i}")

    def extend(self, iterable):
        if len(iterable) <= len([element for element in self.deque if element is None]):
            for element in iterable:
                self.append(element)
        else:
            raise ValueError(f"Not enough space for all elements in {iterable}")

    def extendleft(self, iterable):
        if len(iterable) <= (self.tail - self.head - 1) % self.max_len:
            for element in iterable:
                self.appendleft(element)
        else:
            raise ValueError(f"Not enough space for all elements in {iterable}")

    def index(self, x, start=None, stop=None):
        if start is None:
            start = 0
        if stop is None:
            stop = self.max_len - 1
        if start < 0 or stop >= self.max_len:
            raise ValueError("Invalid indices")
        else:
            for i in range(start, stop + 1):
                if self.deque[i] == x:
                    print(f"Value {x} found at index {i}")
                    break
            else:
                print(f"{x} not found between indices {start} and {stop} ")

    def popleft(self):
        if self.is_empty():
            raise ValueError("Deque is empty")

        self.tail = 0

        while self.deque[self.tail] is None:
            self.tail = (self.tail + 1) % self.max_len

        popped_element = self.deque[self.tail]
        self.deque[self.tail] = None
        self.tail = (self.tail + 1) % self.max_len

        return popped_element

    def pop(self):
        if self.is_empty():
            raise ValueError("Deque is empty")

        self.head = self.max_len - 1

        while self.deque[self.head] is None:
            self.head = (self.head - 1) % self.max_len

        popped_element = self.deque[self.head]
        self.deque[self.head] = None
        self.head = (self.head - 1) % self.max_len

        return popped_element

    def __str__(self):
        if self.deque:
            return " ".join(str(x) for x in self.deque)


static_deque = StaticDeque(5)

print(static_deque.pop())

print(static_deque)

File: exercise/src/coursework/StaticDeque.java
package coursework;

import java.util.Arrays;

class StaticDeque {
    private final int maxLen;
    private Object[] deque;
    private int head;
    private int tail;

    public StaticDeque(int max_len) {
        this.maxLen = max_len;
        this.deque = new Object[max_len];
        this.head = 0;
        this.tail = 0;
    }

    public boolean isFull() {
        return deque[tail] != null;
    }

    public boolean isEmpty() {
        for (Object el : deque) {
            if (el != null) {
                return false;
            }
        }
        return true;
    }

    public String appendLeft(Object x) {
        if (isFull()) {
            return "Full deque, cannot append " + x + " left";
        } else {
            deque[tail] = x;
            tail = (tail + 1) % maxLen;
            return "Element " + x + " added at position " + tail;
        }
    }

    public String append(Object x) {
        if (isFull()) {
            return "Full deque, cannot append " + x;
        } else {
            head = (head - 1 + maxLen) % maxLen;
            deque[head] = x;
            return "Element " + x + " added at position " + head;
        }
    }

    public String clear() {
        if (!Arrays.equals(deque, new Object[maxLen])) {
            head = 0;
            tail = 0;
            deque = new Object[maxLen];
            return "Deque cleared";
        } else {
            return "Deque is already cleared";
        }
    }

    public StaticDeque copy() {
        StaticDeque new_deque = new StaticDeque(maxLen);
        new_deque.deque = Arrays.copyOf(deque, maxLen);
        new_deque.head = head;
        new_deque.tail = tail;
        return new_deque;
    }

    public int count(Object x) {
        int count = 0;
        for (Object el : deque) {
            if (el != null && el.equals(x)) {
                count++;
            }
        }
        return count;
    }

    public void insert(int i, Object x) {
        if (!(0 <= i && i < maxLen)) {
            System.out.println("Index out of range");
        } else if (deque[i] == null) {
            deque[i] = x;
        } else {
            System.out.println("There's already value (" + deque[i] + ") at index " + i);
        }
    }

    public void extend(Object[] iterable) {
        if (iterable.length <= count(null)) {
            for (Object element : iterable) {
                append(element);
            }
        } else {
            throw new IllegalArgumentException("Not enough space for all elements in " + Arrays.toString(iterable));
        }
    }

    public void extendLeft(Object[] iterable) {
        if (!isFull()) {
            for (Object element : iterable) {
                appendLeft(element);
            }
        } else {
            throw new IllegalArgumentException("Not enough space for all elements in " + Arrays.toString(iterable));
        }
    }

    public void index(Object x, Integer start, Integer stop) {
        if (start == null) {
            start = 0;
        }
        if (stop == null) {
            stop = maxLen - 1;
        }
        if (start < 0 || stop >= maxLen) {
            throw new IllegalArgumentException("Invalid indices");
        } else {
            for (int i = start; i <= stop; i++) {
                if (deque[i] != null && deque[i].equals(x)) {
                    System.out.println("Value " + x + " found at index " + i);
                    return;
                }
            }
            System.out.println(x + " not found between indices " + start + " and " + stop);
        }
    }

    public Object popLeft() {
        if (isEmpty()) {
            throw new IllegalArgumentException("Deque is empty");
        }
        tail = 0;
        while (deque[tail] == null) {
            tail = (tail + 1) % maxLen;
        }
        Object popped_element = deque[tail];
        deque[tail] = null;
        tail = (tail + 1) % maxLen;
        return "Removed element: "+ popped_element;
    }

    public Object pop() {
        if (isEmpty()) {
            throw new IllegalArgumentException("Deque is empty");
        }
        head = (head - 1 + maxLen) % maxLen;
        while (deque[head] == null) {
            head = (head - 1 + maxLen) % maxLen;
        }
        Object popped_element = deque[head];
        deque[head] = null;
        head = (head - 1 + maxLen) % maxLen;
        return "Popped " + popped_element;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        for (Object x : deque) {
            sb.append((x != null ? x : "null")).append(" ");
        }
        return "Deque: " + sb.toString().trim();
    }
}


